include/coro/semaphore.hpp

```diff
@@ -167,12 +168,14 @@ class semaphore
      * Stops the semaphore and will notify all release/acquire waiters to wake up in a failed state.
      * Once this is set it cannot be un-done and all future oprations on the semaphore will fail.
      */
-    auto shutdown() noexcept -> void
+    [[nodiscard]] auto shutdown() noexcept -> coro::task<void>
     {
+        auto lock = co_await m_mutex.scoped_lock();
         bool expected{false};
```

Comment:
Should check the current state first so that if we're already shutdown we can return straight away.

Reply:
yeah can probably add it before the lock acquire `if (is_shutdown() { co_return; }`?

Reply:
Done.